### ISSUE TITLE
[5.0] Session Persistance Issues

### ISSUE THREAD

**Issue by GrahamCampbell** [laravel/framework#8172]
#6777 and #5416 still seem to be an issue, even for drivers other than `file`.

---
// cc @ammont, @esbenp, @barryvdh, @bonzai, @henrikromby, @avi123,

**Comment by ammont** [comment]
We are experiencing token regenerating issue on **Redis**, I don't think it would be lock issue because redis is single thread and atomic, I used the method suggested in [#6777](https://github.com/laravel/framework/issues/6777) to debug and here is the log:

```shell
[2015-03-26 13:12:06] production.INFO: ["k9OqsVTnE0QkphMjGLxfF8nnClidNXsDZo7i9H8C","k9OqsVTnE0QkphMjGLxfF8nnClidNXsDZo7i9H8C","1a85f2d0adb856b874dda75f05c75ea8a7ad0c43"] [] []
[2015-03-26 13:12:06] production.INFO: ["k9OqsVTnE0QkphMjGLxfF8nnClidNXsDZo7i9H8C","k9OqsVTnE0QkphMjGLxfF8nnClidNXsDZo7i9H8C","1a85f2d0adb856b874dda75f05c75ea8a7ad0c43"] [] []
[2015-03-26 13:12:06] production.INFO: ["k9OqsVTnE0QkphMjGLxfF8nnClidNXsDZo7i9H8C","k9OqsVTnE0QkphMjGLxfF8nnClidNXsDZo7i9H8C","1a85f2d0adb856b874dda75f05c75ea8a7ad0c43"] [] []
[2015-03-26 13:12:06] production.INFO: ["k9OqsVTnE0QkphMjGLxfF8nnClidNXsDZo7i9H8C","k9OqsVTnE0QkphMjGLxfF8nnClidNXsDZo7i9H8C","1a85f2d0adb856b874dda75f05c75ea8a7ad0c43"] [] []
[2015-03-26 13:12:06] production.INFO: ["k9OqsVTnE0QkphMjGLxfF8nnClidNXsDZo7i9H8C","k9OqsVTnE0QkphMjGLxfF8nnClidNXsDZo7i9H8C","1a85f2d0adb856b874dda75f05c75ea8a7ad0c43"] [] []
[2015-03-26 13:12:06] production.INFO: ["k9OqsVTnE0QkphMjGLxfF8nnClidNXsDZo7i9H8C","k9OqsVTnE0QkphMjGLxfF8nnClidNXsDZo7i9H8C","1a85f2d0adb856b874dda75f05c75ea8a7ad0c43"] [] []
[2015-03-26 13:12:07] production.INFO: ["k9OqsVTnE0QkphMjGLxfF8nnClidNXsDZo7i9H8C","k9OqsVTnE0QkphMjGLxfF8nnClidNXsDZo7i9H8C","1a85f2d0adb856b874dda75f05c75ea8a7ad0c43"] [] []
[2015-03-26 13:12:07] production.INFO: ["k9OqsVTnE0QkphMjGLxfF8nnClidNXsDZo7i9H8C","k9OqsVTnE0QkphMjGLxfF8nnClidNXsDZo7i9H8C","1a85f2d0adb856b874dda75f05c75ea8a7ad0c43"] [] []
[2015-03-26 13:12:07] production.INFO: ["k9OqsVTnE0QkphMjGLxfF8nnClidNXsDZo7i9H8C","k9OqsVTnE0QkphMjGLxfF8nnClidNXsDZo7i9H8C","1a85f2d0adb856b874dda75f05c75ea8a7ad0c43"] [] []
[2015-03-26 13:12:07] production.INFO: ["k9OqsVTnE0QkphMjGLxfF8nnClidNXsDZo7i9H8C","k9OqsVTnE0QkphMjGLxfF8nnClidNXsDZo7i9H8C","1a85f2d0adb856b874dda75f05c75ea8a7ad0c43"] [] []
[2015-03-26 13:12:07] production.INFO: ["k9OqsVTnE0QkphMjGLxfF8nnClidNXsDZo7i9H8C","k9OqsVTnE0QkphMjGLxfF8nnClidNXsDZo7i9H8C","1a85f2d0adb856b874dda75f05c75ea8a7ad0c43"] [] []
[2015-03-26 13:12:08] production.INFO: ["k9OqsVTnE0QkphMjGLxfF8nnClidNXsDZo7i9H8C","k9OqsVTnE0QkphMjGLxfF8nnClidNXsDZo7i9H8C","1a85f2d0adb856b874dda75f05c75ea8a7ad0c43"] [] []
[2015-03-26 13:12:08] production.INFO: ["k9OqsVTnE0QkphMjGLxfF8nnClidNXsDZo7i9H8C","9uIpGse9rdQlZJGs8NYMFH4s8KsWkvDsZ2efk5Fv","1a85f2d0adb856b874dda75f05c75ea8a7ad0c43"] [] []
[2015-03-26 13:12:08] production.INFO: ["k9OqsVTnE0QkphMjGLxfF8nnClidNXsDZo7i9H8C","9uIpGse9rdQlZJGs8NYMFH4s8KsWkvDsZ2efk5Fv","1a85f2d0adb856b874dda75f05c75ea8a7ad0c43"] [] []
[2015-03-26 13:12:08] production.INFO: ["k9OqsVTnE0QkphMjGLxfF8nnClidNXsDZo7i9H8C","9uIpGse9rdQlZJGs8NYMFH4s8KsWkvDsZ2efk5Fv","1a85f2d0adb856b874dda75f05c75ea8a7ad0c43"] [] []
[2015-03-26 13:12:08] production.INFO: ["k9OqsVTnE0QkphMjGLxfF8nnClidNXsDZo7i9H8C","9uIpGse9rdQlZJGs8NYMFH4s8KsWkvDsZ2efk5Fv","1a85f2d0adb856b874dda75f05c75ea8a7ad0c43"] [] []
[2015-03-26 13:12:09] production.INFO: ["k9OqsVTnE0QkphMjGLxfF8nnClidNXsDZo7i9H8C","9uIpGse9rdQlZJGs8NYMFH4s8KsWkvDsZ2efk5Fv","1a85f2d0adb856b874dda75f05c75ea8a7ad0c43"] [] []
[2015-03-26 13:12:09] production.INFO: ["k9OqsVTnE0QkphMjGLxfF8nnClidNXsDZo7i9H8C","9uIpGse9rdQlZJGs8NYMFH4s8KsWkvDsZ2efk5Fv","1a85f2d0adb856b874dda75f05c75ea8a7ad0c43"] [] []
[2015-03-26 13:12:10] production.INFO: ["k9OqsVTnE0QkphMjGLxfF8nnClidNXsDZo7i9H8C","9uIpGse9rdQlZJGs8NYMFH4s8KsWkvDsZ2efk5Fv","1a85f2d0adb856b874dda75f05c75ea8a7ad0c43"] [] []
[2015-03-26 13:12:10] production.INFO: ["k9OqsVTnE0QkphMjGLxfF8nnClidNXsDZo7i9H8C","9uIpGse9rdQlZJGs8NYMFH4s8KsWkvDsZ2efk5Fv","1a85f2d0adb856b874dda75f05c75ea8a7ad0c43"] [] []
[2015-03-26 13:12:10] production.INFO: ["k9OqsVTnE0QkphMjGLxfF8nnClidNXsDZo7i9H8C","9uIpGse9rdQlZJGs8NYMFH4s8KsWkvDsZ2efk5Fv","1a85f2d0adb856b874dda75f05c75ea8a7ad0c43"] [] []
[2015-03-26 13:12:10] production.INFO: ["k9OqsVTnE0QkphMjGLxfF8nnClidNXsDZo7i9H8C","9uIpGse9rdQlZJGs8NYMFH4s8KsWkvDsZ2efk5Fv","1a85f2d0adb856b874dda75f05c75ea8a7ad0c43"] [] []
[2015-03-26 13:12:10] production.INFO: ["k9OqsVTnE0QkphMjGLxfF8nnClidNXsDZo7i9H8C","9uIpGse9rdQlZJGs8NYMFH4s8KsWkvDsZ2efk5Fv","1a85f2d0adb856b874dda75f05c75ea8a7ad0c43"] [] []
[2015-03-26 13:12:11] production.INFO: ["k9OqsVTnE0QkphMjGLxfF8nnClidNXsDZo7i9H8C","9uIpGse9rdQlZJGs8NYMFH4s8KsWkvDsZ2efk5Fv","1a85f2d0adb856b874dda75f05c75ea8a7ad0c43"] [] []
[2015-03-26 13:12:11] production.INFO: ["k9OqsVTnE0QkphMjGLxfF8nnClidNXsDZo7i9H8C","9uIpGse9rdQlZJGs8NYMFH4s8KsWkvDsZ2efk5Fv","1a85f2d0adb856b874dda75f05c75ea8a7ad0c43"] [] []
[2015-03-26 13:12:11] production.INFO: ["k9OqsVTnE0QkphMjGLxfF8nnClidNXsDZo7i9H8C","9uIpGse9rdQlZJGs8NYMFH4s8KsWkvDsZ2efk5Fv","1a85f2d0adb856b874dda75f05c75ea8a7ad0c43"] [] []
[2015-03-26 13:12:11] production.INFO: ["k9OqsVTnE0QkphMjGLxfF8nnClidNXsDZo7i9H8C","9uIpGse9rdQlZJGs8NYMFH4s8KsWkvDsZ2efk5Fv","1a85f2d0adb856b874dda75f05c75ea8a7ad0c43"] [] []
[2015-03-26 13:12:11] production.INFO: ["k9OqsVTnE0QkphMjGLxfF8nnClidNXsDZo7i9H8C","9uIpGse9rdQlZJGs8NYMFH4s8KsWkvDsZ2efk5Fv","1a85f2d0adb856b874dda75f05c75ea8a7ad0c43"] [] []
[2015-03-26 13:12:11] production.INFO: ["k9OqsVTnE0QkphMjGLxfF8nnClidNXsDZo7i9H8C","9uIpGse9rdQlZJGs8NYMFH4s8KsWkvDsZ2efk5Fv","1a85f2d0adb856b874dda75f05c75ea8a7ad0c43"] [] []
```

you can see that the token is regenerated out of nowhere.

**one more thing**, when I was testing the application (sending many concurrent requests) on another tab I was not able to login!

**Comment by ammont** [comment]
maybe related to this said by @avi123 in [#5416](https://github.com/laravel/framework/issues/5416#issuecomment-68395469):
>The first issue occurs if you have multiple concurrent POSTs/PUTs, or even sequential asynchronous requests that happen in close succession such that the second request starts prior to the first request completing. The issue here is that changes to the session in the second request can be lost if the first request doesn't lock the session, but instead does a read, and then a full write at the completion of the request (I believe the method that the Session Store takes). This is the impetus for using a locking session driver in general.

**Comment by ifox** [comment]
I can confirm this is happening using the file driver as well as the memcached driver.

**Comment by ammont** [comment]
Ok I added some code to log when a session starts when it stops (saves) and when the cookie is sent and here is the result:
```
[2015-04-07 15:11:33] local.NOTICE: CLOSED: c9dab98e17844c7127b47cfc28429fdc57ee4d9b [] []
[2015-04-07 15:11:33] local.NOTICE: COOKIE SENT: laravel_session=c9dab98e17844c7127b47cfc28429fdc57ee4d9b; expires=Tue, 07-Apr-2015 17:11:33 GMT; path=/; domain=.adview.local; httponly [] []
[2015-04-07 15:11:33] local.NOTICE: CLOSED: c9dab98e17844c7127b47cfc28429fdc57ee4d9b [] []
[2015-04-07 15:11:33] local.NOTICE: COOKIE SENT: laravel_session=c9dab98e17844c7127b47cfc28429fdc57ee4d9b; expires=Tue, 07-Apr-2015 17:11:33 GMT; path=/; domain=.adview.local; httponly [] []
[2015-04-07 15:11:33] local.INFO: session id regenerated: 3fae6afa0b40226cfba997b9ace82e73127c4791 [] []
[2015-04-07 15:11:33] local.INFO: ["EcSKhVhL8QbNMhmgrJp7nAIeXe1Z3ThuL2T8bRmx","EcSKhVhL8QbNMhmgrJp7nAIeXe1Z3ThuL2T8bRmx","c9dab98e17844c7127b47cfc28429fdc57ee4d9b"] [] []
[2015-04-07 15:11:33] local.NOTICE: CLOSED: c9dab98e17844c7127b47cfc28429fdc57ee4d9b [] []
[2015-04-07 15:11:33] local.NOTICE: COOKIE SENT: laravel_session=c9dab98e17844c7127b47cfc28429fdc57ee4d9b; expires=Tue, 07-Apr-2015 17:11:33 GMT; path=/; domain=.adview.local; httponly [] []
[2015-04-07 15:11:33] local.NOTICE: CLOSED: c9dab98e17844c7127b47cfc28429fdc57ee4d9b [] []
[2015-04-07 15:11:33] local.NOTICE: COOKIE SENT: laravel_session=c9dab98e17844c7127b47cfc28429fdc57ee4d9b; expires=Tue, 07-Apr-2015 17:11:33 GMT; path=/; domain=.adview.local; httponly [] []
[2015-04-07 15:11:33] local.NOTICE: CLOSED: 3fae6afa0b40226cfba997b9ace82e73127c4791 [] []
[2015-04-07 15:11:33] local.NOTICE: COOKIE SENT: laravel_session=3fae6afa0b40226cfba997b9ace82e73127c4791; expires=Tue, 07-Apr-2015 17:11:33 GMT; path=/; domain=.adview.local; httponly [] []
[2015-04-07 15:11:34] local.INFO: ["EcSKhVhL8QbNMhmgrJp7nAIeXe1Z3ThuL2T8bRmx","EcSKhVhL8QbNMhmgrJp7nAIeXe1Z3ThuL2T8bRmx","c9dab98e17844c7127b47cfc28429fdc57ee4d9b"] [] []
[2015-04-07 15:11:34] local.NOTICE: CLOSED: c9dab98e17844c7127b47cfc28429fdc57ee4d9b [] []
[2015-04-07 15:11:34] local.NOTICE: COOKIE SENT: laravel_session=c9dab98e17844c7127b47cfc28429fdc57ee4d9b; expires=Tue, 07-Apr-2015 17:11:34 
```
as you can see in the middle the session id is regenerated because I have tried logging in when sending concurrent requests, and as you can see while the session id is changed the other concurrent requests are using the prior id so on next response the cookie is rewrote and you're logged out. any idea what can make this happen?

**Comment by ifox** [comment]
After testing on multiple machines / configurations, I was able to notice that it happens only using homestead without nfs folder sync.

**Comment by ammont** [comment]
@GrahamCampbell why is this closed?

**Comment by TheShiftExchange** [comment]
I tested this on Homestead v2.5 on my Windows 7 64bit box using VMWare. I tried using session drivers `file`, `redis` and `memcache`.

I could not replicate despite doing thousands of AJAX requests.

I'll keep trying some other methods...

**Comment by GrahamCampbell** [comment]
I Taylor and I can't replicate this either, but it seems there are a good number of people that seem to have having this issue.

**Comment by Naxiz** [comment]
Could it have something to do with a change or bug in the latest PHP version?

I'm using PHP 5.6.7 for both local development and my servers, and I have this issue on both. The rest of the setup is totally different (Windows/Debian, PHP server/nginx).

**Comment by TheShiftExchange** [comment]
@Naxiz - the issue has been known since at least Dec 2014. So it was earlier than PHP 5.6.4

But there must be some combination of PHP/Server/Laravel versions that is causing it. Or perhaps even the OS or Server.

Perhaps if more people with the issue can post their system configs/versions - we might see a common setup between them that we can use to narrow down the issue?

**Comment by TheShiftExchange** [comment]
I've tried further on a new Forge DigitalOcean box (lowest box - 512mb ram) - running PHP 5.6.7.

I also tried with browsers Chrome41 + IE11

Still cant replicate :(

**Comment by Sladewill** [comment]
Its happening for me under xampp 3.2.1, windows 7, php 5.6, mysql 5.5.34, apache 2.4.7
As well on windows 8.1, php 5.6, mysql 5.6.24, nginx 1.7.12

**Comment by TheShiftExchange** [comment]
@Sladewill @Naxiz - what browsers are you testing with? (I know it shouldnt matter - but I'm trying to think of every possibility)

**Comment by Sladewill** [comment]
Firefox and Chrome.

**Comment by TheShiftExchange** [comment]
@Sladewill - which exact browser versions are you using?

And what code are you running to cause this issue? If you run the code in https://github.com/laravel/framework/issues/6777 do you get the problem?

**Comment by Sladewill** [comment]
Home Pc
Firefox 37.0.1
Chrome 41.0.2272.118 m

Work PC
Firefox Developer 39.0a2 (2015-04-06)
Chrome 41.0.2272.118 m

I'll test the code when I get home.
It happens very inconsistently with both the general Auth + Sentinel.

**Comment by Naxiz** [comment]
My PC is running Windows 8.1, PHP 5.6.7, the PHP web server and Chrome 41.0.2272.118 m (64-bit). My first server is running Debian Sid, PHP 5.6.7 and nginx 1.6.2. My second server is running Debian Jessie, PHP 5.6.7 and nginx 1.6.2 also.

@TheShiftExchange I've tried the code of issue #6777. I've run ~2500 queries without any mismatch (I checked the logs). I've also created a new Laravel 5 project and copied the added code into it, but can't reproduce it there either. Using my own Naxiz/L5-CSRF-TestCase@74d8222551cd0383c5887b108967b634c7af3b15 seconds after it, I get an exception after just 2 clicks.

**Comment by esbenp** [comment]
The #6777 was created on my local OSX dev machine using homestead (don't remember the version though). We initially found the problem on our DigitalOcean servers created through Forge.

**Comment by ammont** [comment]
@taylorotwell @GrahamCampbell I can easily replicate this error on our server and on my local machine, they have different versions of php different os, etc., I use laravel 4.1, on login form I simply click or enter the submit button 20-30 times repeatedly fast. as you know even on login the csrf token does not change, but I get a token mismatch error on this situation.

**Comment by TheShiftExchange** [comment]
@ammont @esbenp @Naxiz - can you visit this server - and tell me if you can make it fail: http://45.55.164.181/

This is using the exact fork of https://github.com/Naxiz/L5-CSRF-TestCase/commit/74d8222551cd0383c5887b108967b634c7af3b15 - on a DigitalOcean server using Forge.

I still cant replicate - but I'm wondering if you can make it fail?

**Comment by jhauraw** [comment]
We've been battling with this puzzle for 2 months now. On our busy production system, we get 2-3 `TokenMismatchExceptions` per 24-hr period upon a simple form submission. During this same period we have ~150 succeed without issue.

We've implemented very detailed debug logging and one pattern has emerged, which is browser related. Over the last 2 months, with 2-3 issues per-day, only these browsers have triggered the issue (by quantity, predominantly IE 11 and Android 4.4.* Chrome 41):

```
IE 11, Windows 8.1 - Mozilla/5.0 (Windows NT 6.3; WOW64; Trident/7.0; rv:11.0) like Gecko

WebViewer 4.0, Android 4.4.2 - Mozilla/5.0 (Linux; U; Android 4.4.2; en-us; LGLS620 Build/KOT49I.LS620ZV3) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/30.0.1599.103 Mobile Safari/537.36

Chrome 41, Android 4.4.4 - Mozilla/5.0 (Linux; Android 4.4.4; SAMSUNG-SM-G870A Build/KTU84P) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/41.0.2272.96 Mobile Safari/537.36

Chrome 41, Android 4.4.2 - Mozilla/5.0 (Linux; Android 4.4.2; SCH-I545 Build/KOT49H) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/41.0.2272.96 Mobile Safari/537.36

Safari 8.0, IOS 8.2 - Mozilla/5.0 (iPhone; CPU iPhone OS 8_2 like Mac OS X) AppleWebKit/600.1.4 (KHTML, like Gecko) Version/8.0 Mobile/12D508 Safari/600.1.4
```

Related SO question:
http://stackoverflow.com/questions/29380806/tokenmismatch-only-certain-browsers-laravel-5-fresh-production

We have a Forge managed Linode with LEMP (nginx 1.6.2, php 5.6.6-1+deb.sury.org~trusty+1 and Laravel 5 up-to-date).

So far, we have been unable to replicate in our testing on the said browsers.

Here's an example log run for IE 10:

```
// ORIGINAL SESSION STARTED
Apr 12 13:37:23 Session {"start":1428871043,"id":"bd5181ecfada5504c7f1b07522b3c25f94ca7a0f","token":"dh7iw7CR3rlXcvLd51zSQH79PWgAkH2fyOs6cKBl","ip":"173.79.XX.XXX","method":"GET","url":"https://sub.domain.com/place/name","ua":"Mozilla/5.0 (compatible; MSIE 10.0; Windows NT 6.2; WOW64; Trident/6.0)"} []

// FORM SUBMITTED 2.5 minutes later.
Apr 12 13:41:00 SESSION ID: 461bb378850caac38bd8881979e866dddf608113 [] []
Apr 12 13:41:00 SESSION token: Df0vWqhKHbzIdt4nwbhse5htywhMRUkyPl7p0lRs [] []
Apr 12 13:41:00 REQUEST INPUT _token: dh7iw7CR3rlXcvLd51zSQH79PWgAkH2fyOs6cKBl [] []
Apr 12 13:41:00 REQUEST HEADER X-CSRF-TOKEN:  [] []
Apr 12 13:41:00 REQUEST HEADER X-XSRF-TOKEN:  [] []
Apr 12 13:41:00 REQUEST URL: https://sub.domain.com/requests/post [] []

// RAW REQUEST
Apr 12 13:41:00 array (   'USER' => 'forge',   'HOME' => '/home/forge',   'FCGI_ROLE' => 'RESPONDER',   'APP_ENV' => 'production',   'QUERY_STRING' => '',   'REQUEST_METHOD' => 'POST',   'CONTENT_TYPE' => 'application/x-www-form-urlencoded',   'CONTENT_LENGTH' => '275',   'SCRIPT_FILENAME' => '/home/forge/domain.com/public/index.php',   'SCRIPT_NAME' => '/index.php',   'REQUEST_URI' => '/requests/post',   'DOCUMENT_URI' => '/index.php',   'DOCUMENT_ROOT' => '/home/forge/domain.com/public',   'SERVER_PROTOCOL' => 'HTTP/1.1',   'GATEWAY_INTERFACE' => 'CGI/1.1',   'SERVER_SOFTWARE' => 'nginx/1.6.2',   'REMOTE_ADDR' => '173.79.XX.XXX',   'REMOTE_PORT' => '62854',   'SERVER_ADDR' => '104.237.XX.XXX',   'SERVER_PORT' => '443',   'SERVER_NAME' => 'domain.com',   'HTTPS' => 'on',   'REDIRECT_STATUS' => '200',   'HTTP_ACCEPT' => 'text/html, application/xhtml+xml, */*',   'HTTP_REFERER' => 'https://sub.domain.com/place/name?t=%2Bkeyword',   'HTTP_ACCEPT_LANGUAGE' => 'en-US',   'HTTP_USER_AGENT' => 'Mozilla/5.0 (compatible; MSIE 10.0; Windows NT 6.2; WOW64; Trident/6.0)',   'HTTP_CONTENT_TYPE' => 'application/x-www-form-urlencoded',   'HTTP_ACCEPT_ENCODING' => 'gzip, deflate',   'HTTP_HOST' => 'sub.domain.com',   'HTTP_CONTENT_LENGTH' => '275',   'HTTP_DNT' => '1',   'HTTP_CONNECTION' => 'Keep-Alive',   'HTTP_CACHE_CONTROL' => 'no-cache',   'HTTP_COOKIE' => '_ga=GA1.2.1070026939.1428881850; _gat=1',   'PHP_SELF' => '/index.php',   'REQUEST_TIME_FLOAT' => 1428871260.0079651,   'REQUEST_TIME' => 1428871260, ) [] []

Apr 12 13:41:00 REQUEST PATH: requests/post [] []
Apr 12 13:41:00 SERVER: array (   '_ga' => NULL,   '_gat' => NULL, ) [] []
Apr 12 13:41:00 COOKIE: array (   '_token' => 'Df0vWqhKHbzIdt4nwbhse5htywhMRUkyPl7p0lRs', ) [] []

Apr 12 13:41:00 TokenMismatchExpection thrown

Apr 12 13:41:00 ExceptionHandler@render: TokenMismatchException: Redirecting back. {"previous":null} []

// 2nd SESSION STARTED
Apr 12 13:41:00 Session {"start":1428871260,"id":"f408a809ca246322928880a47509a8d0b455e519","token":"aPWGA6oOn13oayiVez0VhGuG2SYajD6wkr7wFfMy","ip":"173.79.XX.XXX","method":"GET","url":"https://sub.domain.com/place/name","ua":"Mozilla/5.0 (compatible; MSIE 10.0; Windows NT 6.2; WOW64; Trident/6.0)"} []
```

**Comment by TheShiftExchange** [comment]
Ping @ammont @esbenp @Naxiz @Sladewill.

Can you please see if you have the problem on this server:  http://45.55.164.181/

I cant replicate - but I'm wondering if you can? Might help pinpoint the issue.

**Comment by Sladewill** [comment]
As far as I can see on that server specifically the issue doesn't seem to occur. But another issue that could be related is the fact the laravel session cookie seems to expire randomly, when it does a new session gets created like when you logout so you would lose that token?

**Comment by ammont** [comment]
@taylorotwell @GrahamCampbell @TheShiftExchange @esbenp @Naxiz @Sladewill 
After a lot of investigation on the issue in our company, I came to this result, I hope it helps, first of all here is our environment specifications:

PHP: 5.3.3
LARAVEL: 4.1
OS: centos 6 on server and os x mavericks in development environment
APACHE: 2
MYSQL: 5.6.19
REDIS: 2.4.10
PREDIS:  0.8.*

First of all it seems that the TokenMistmatch exception occurs in a varied different conditions, I nearly investigated all of them and was able to solve some of them, some depends on the logic behind the session and some can be bugs. In the following i will explain each situation that I have faced.

**1. Expired sessions**
Let's say you have configured your session for 3 hours a user opens up a form and for some reason he leaves the computer (getting a cup of coffee) so after 3 hours when the session is expired he tries to submit a form and gets a token exception. this is why everyone once in while gets a token errror regardless of how much stable the application is, I can imagine of 2 ways to prevent it and they're renewing the session cookie using ajax on a timely basis, or increasing the session expire time to a considerable amount of time.

**2. Concurrent requests when session is expired**
Sometimes on the load of your first page concurrent requests happen, for example the user has two different tabs open on chrome and when he reopens chrome chrome sends the requests simultaneously or you may have multiple concurrent ajax request on the load of the first page of your application. so note that the session cookie is received after the first response is received but you may send the next request before this happens and therefore you will get a new session (new cookie) on each requests, this can lead to token exception if one of these requests populates a form. you can prevent this scenario based on it's source, for example if the ajax request are causing the problem and you do not use session in ajax responses, you can disable sending the session cookie if the request is ajax, in the second scenario (clicking the submit button multiple times) you can simple disable the button once it's submitted.


**3. Concurrent requests on login**
When login is occurred, laravel for security reasons changes the session id, copy the session data and DESTROYS THE LAST SESSION so let say for some reason when logging in concurrent requests happen (clicking the login button multiple times) so the session id would be regenerated multiple times and DESTROYS the last generated sessions in the server side, as some of these requests still use the prior session id (which does not exist in the server side anymore) it leads to regenerating the CSRF token, please note that normally laravel does not regenerate the token on login if it can find the token in guest (not logged in) session, but in this case as the guest session is destroyed as a result it will regenerate the token and it can result to token exception in other requests that are using the original token. also note that not only this issue can result in token exception it can also result in the user being logged out after one request, because the concurrent requests can change the logged in session. I solved this issue by changing one line in the laravel code, in `Illuminate\Auth\Guard:updateSession`:

```php
protected function updateSession($id)
{
    $this->session->put($this->getName(), $id);

    //$this->session->migrate(true);
    $this->session->migrate()
}
```
passing true to the migrate method of the session store will result in destroying the session on the server after migration, so now the data will remain on the server and destroyed on their expire time rather than on this request and it will solve the issue. I don’t know if we can call this a bug in laravel, but I guess we can come up with a better solution for this.

**4. Browsers**
After investigating the logs it turned out that some of these token exceptions were because of the user’s browser not accepting the session cookie,I saw it the most on iOS Safari, I believe this is something we can do nothing about it.

**5. Redis bug**
Some of the token exceptions on our sever was because of redis, for some reason laravel could not read the session data from the redis server when opening the session so it would result to the regeneration of the CSRF token. it was happening randomly on our server, so I tried changing the session driver and this type of exceptions faded away. I tried database, apc and file drivers and none produced this issue. I have not yet found what is causing the bug but i think it can be a bug with redis or the predis library. (as you know laravel 4.1 is not using the latest version of predis because of compatibility issues).

Okay these are my experiences of the last 2 month working on this issue. I hope it may change your point of view regarding solution to this issue, the most important thing is that the token exception does not happen because of one reason it’s can be a result of multiple issue. please share with me if you have had similar incident or you have something new.

**Comment by GrahamCampbell** [comment]
> LARAVEL: 4.1

We're not interested in issues with 4.1.

> PHP: 5.3.3

We're not interested with issues with PHP 5.3.

**Comment by GrahamCampbell** [comment]
We want replication on Laravel 5.x.

**Comment by GrahamCampbell** [comment]
> We're not interested in issues with 4.1.

This is because it doesn't pose a security issue so we have no intention to patch it on 4.1.x.

**Comment by ammont** [comment]
@GrahamCampbell I'm aware but I was trying to lighten things up, wondering if anyone can test the similar situations on 5, I'm also upgrading the laravel of my application so then I can try again with L5.

**Comment by GrahamCampbell** [comment]
@ammont Thank you. :)

**Comment by Naxiz** [comment]
@TheShiftExchange I can't reproduce it on your server. I also set up a new server at DigitalOcean and can't reproduce it there either.

I thought it may have something to do with the time zone, because it's Europe/Amsterdam on all machines I can reproduce the issue on, but not on DigitalOcean, but after changing it I still can't reproduce it.

**Comment by sn0opr** [comment]
I got the same issue, I get some data that appear and desappears randomly, I'm guessing that it's from  the ajax long polling in my app.

**Comment by ammont** [comment]
@sn0opr Can you please provide your laravel version, php version, session driver and the driver version and the exact way you are able to replicate this issue (even with sample code). thanks.

**Comment by yoramdelangen** [comment]
Im not sure if this related to this issue, but my session variables are not saved in the file session. I tried to quickly reproduce it with the following code in the latest Laravel (5.1-dev): 

```php
// routes.php
Route::get('test', function() {
    return view('test');
});

Route::post('test', function() {
    $setErrors = [
        'error 1' => 'This is my error test 1',
        'error 2' => 'Error num2',
    ];
    return redirect()->to('test')->withErrors($setErrors);
});

// test.blade.php
{{ print_r($errors, true)  }}
<hr>
{{ print_r($_SERVER['REQUEST_METHOD'], true) }}

{!! Form::open(['url' => 'test', 'method' => 'post']) !!}
{!! Form::submit('Submit', ['class' => 'form-control']) !!}
{!! Form::close() !!}
```

In the example i except to get a filled `Illuminate\Support\ViewErrorBag` with 2 messages, but unfortunately i do not. When i try this exact example within Laravel 5.0.22 its working properly. 

After some debugging and digging within the `terminateMiddleware` it should save all attributes (with `Illuminate\Session\Middleware\StartSession::terminate()`). But somehow the function is not triggered, but the if-loop requirements are all valid.

```php
// terminate function within startSession
public function terminate($request, $response)
{
	if ($this->sessionHandled && $this->sessionConfigured() && ! $this->usingCookieSessions())
	{
		$this->manager->driver()->save();
	}
}
``` 

> Note, ive been following the full stacktrace for the request and until the end the attributes from the session are correct, but not saved in the session file.

**Comment by sn0opr** [comment]
@ammont 

Laravel Framework version 4.2.17

PHP Version 5.5.12

Session driver 'file'

This is a sample code of my long polling method in a regular Controller
```php
public function longPolling($entityId) {
	session_write_close();

	//$someData is an array of key => value data
	if(!($someData = \Session::get('data.'.$entityId))) {
		return \Response::json(['error' => true, ....]);
	} else {
		// Long polling for 30 sec to prevent php execution time error
		$duration = 0; //doing a database query every 3 sec using sleep(3)
		while($duration < 10) {
			//comparing the session variable $someData with database data
			if (/*Some condition...*/) {
				return \Response::json(['update' => true, ....]);
			} else {
				$duration++;
            	sleep(3);
			}
		}
		//if we passed more than 30 sec without new updates,so and the client re-request do another request for another long polling
		return eturn \Response::json(['update' => false]);;
	}
}
```

**Comment by GrahamCampbell** [comment]
> Laravel Framework version 4.2.17

Laravel 4.2 is not supported.

**Comment by TheShiftExchange** [comment]
> Laravel 4.2 is not supported.

Yes - but this bug first appeared in 4.2 (or at least it was first reported then) - and has carried on into 5.0 - so it is still relevant to know of issues in 4.2.

This bug is so elusive - yet it is clearly out there (with some many different server configs etc etc) - that the more info we get the better. We need as many reports as we can get.

And knowing the bug exists in 4.2 helps to drill down the issue. There is obviously something common between 4.2 and 5.0 causing the bug?

**Comment by sn0opr** [comment]
@GrahamCampbell as @TheShiftExchange said the problem is common with v5.0, and what about applications running under Laravel 4.2 like my application which is production.

**Comment by GrahamCampbell** [comment]
> and what about applications running under Laravel 4.2 like my application which is production

We are not supporting 4.2. If we do finally fix this, we will only be merging the fix into 5.0 or 5.1, depending in which one is stable at the time.

**Comment by l3ehnam** [comment]
So I hope people can help to find the bug asap otherwise we all have to migrate to L6.0 or 7.0 because of this bug!

**Comment by radmen** [comment]
I was able to reproduce this bug.

Laravel: 5.0
PHP: 5.6 (provisioned by Forge)
Session drivers: redis, file

The race condition occurred due to images pointing to dynamic URL (image generation was handled by Laravel). Browser was processing the request, but also trying to load the images. The main request was updating the session, but on some cases the requests to images were resetting it.

Consider this simple timeline:

R1: (**SS**)-------------(**DP**)-----------(**SC**)
R2: --------------(**SS**)---------------------------(**SC**)

Legend:
**SS** - session starts - data is read from the storage
**DP** - data is put to the session
**SC** - session is closed - data is written to the storage
**R1** - first request
**R2** - second request

Since the second request has the old data it's writing them back to storage.

I see two solutions for this:

1. simple and naive - save session data only when they were changed
2. session locking

In my case I've simply disabled session saving for image requests.

What do you think about it?

**Comment by GrahamCampbell** [comment]
> simple and naive - save session data only when they were changed

That seems like a good optimisation to make anyway, though if people are storing large objects in the session, could be slow to check for changes.

**Comment by GrahamCampbell** [comment]
@taylorotwell If we make that optimisation, it might be an idea to ditch the serialise/unserialise stuff and not support putting objects in the sessions. It seems like a silly idea to do that anyway?

**Comment by taylorotwell** [comment]
To check for changes you could md5 the serialized array.

**Comment by radmen** [comment]
I had another idea for this - make a flag (e.g `sessionChanged = false` :) ), every call to `put()`, `push()` etc will mark it to `TRUE`. If so - session should be saved.

Does this make sense?

**Comment by GrahamCampbell** [comment]
> To check for changes you could md5 the serialized array.

Yeh, I guess. I was hoping for an optimisation where we could break out before we even serialised anything, since the serialisation process could actually take longer than storing the session in some cases.

**Comment by GrahamCampbell** [comment]
> I had another idea for this - make a flag (e.g sessionChanged = false :) ), every call to put(), push() etc will mark it to TRUE. If so - session should be saved.

That wouldn't work for objects. People can modify them by reference, so they never hit the put or push methods.

**Comment by barryvdh** [comment]
Maybe relevant: http://symfony.com/doc/current/cookbook/session/limit_metadata_writes.html

That references the [WriteCheckSessionHandler](https://github.com/symfony/symfony/blob/2.6/src/Symfony/Component/HttpFoundation/Session/Storage/Handler/WriteCheckSessionHandler.php) and a note:

> Be aware that if the session is not written at every request, it may be garbage collected sooner than usual. This means that your users may be logged out sooner than expected.

**Comment by GrahamCampbell** [comment]
> Be aware that if the session is not written at every request, it may be garbage collected sooner than usual. This means that your users may be logged out sooner than expected.

Hmm, that's a very important point.

**Comment by TheShiftExchange** [comment]
There is this report on SO: http://stackoverflow.com/q/30106131/1317935

'Artisan's cache:clear Flushes User Sessions. Is this Intended?'

It affects Redis according to the post. Might be related? i.e. we've been searching for session issues - but might be a related cache issue as well?

**Comment by GrahamCampbell** [comment]
> "Artisan's cache:clear Flushes User Sessions. Is this Intended?"

Yes. If you're using the same store for sessions as for cache, expect clearing the cache to destroy ALL data in the store.

**Comment by deefour** [comment]
@TheShiftExchange I'm the one who posted that (now deleted) question on SO. I was just looking for the confirmation Graham provided above. It wasn't related to the issue you guys are discussing here.

**Comment by shehi** [comment]
Yea happening to me as well. Same browser, two different environments: one staging server, one my local. Everything works fine on local, however staging environment writes new session to file at every request. Both environments are Ubuntu 15, with latest servers and PHP-FPMs installed.

From time to time my local messes up as well. But miraculously it fixes itself. None of the environments have more than 1 request per 10 secs :) So can't be a load/race-condition issue.

**Comment by shehi** [comment]
Okkk... Right now, I have this problem in both environments and I decided to dig in with XDebug. I will be posting my discoveries here.

Note: I am testing the case in both Chrome and Firefox (latest versions as of now). I have Firebug's latest beta version installed with Firefox to get latest features such as CSS-Maps etc. I also have my app under ```localhost``` and ```audith.basis``` domains set locally of course. I have set ```SESSION_DRIVER=file```. I cleared all my cookies from both browsers before starting the tests and ran commands ```artisan clear:cache``` + ```artisan clear-compiled``` + ```artisan optimize```.

Currently my observations are as following:
1. Chrome console only shows ```XSRF_TOKEN``` cookie only. No ```laravel_session``` cookies whatsoever. Firefox on other hand shows both ```XSRF_TOKEN``` and ```laravel_session``` cookies. Both browsers renew these cookies and write new session file everytime I reload the page.

![image](https://cloud.githubusercontent.com/assets/653569/7678558/ce3a2c02-fd5d-11e4-8a9e-33c3b4a9684c.png)
![image](https://cloud.githubusercontent.com/assets/653569/7678561/d86f8c9e-fd5d-11e4-88ab-4c9bd0611f8c.png)

2. When debugged with XDebug, for line: 
https://github.com/laravel/framework/blob/5.0/src/Illuminate/Session/Middleware/StartSession.php#L118
Firefox actually detects no ```laravel_session``` cookies:
![image](https://cloud.githubusercontent.com/assets/653569/7678607/2d1453f6-fd5e-11e4-91c3-ebd0b3e9259e.png) 

as a value to ```$request->cookies->parameters``` there. Same goes for Chrome as well:
![image](https://cloud.githubusercontent.com/assets/653569/7678654/878fbd5c-fd5e-11e4-8a4f-f914008a52b1.png)

So I guess, none of the browsers actually detect ```laravel_session``` cookie. And if that cookie is not present, new session cookie will be written. Am I missing anything so far or doing something wrong, @GrahamCampbell  @taylorotwell ?

**Comment by shehi** [comment]
Additionally, I gathered: cookies are not allowed to be set in top-level-domains (such as ```localhost```) in Chrome - and from specs-viewpoint, that's the case. Firefox on the other hand, allows setting them for TLDs. Although my code is behind ```audith.basis``` domain, which isn't a TLD, I will setup some bogus real-looking domain and test it again.

**Comment by shehi** [comment]
@GrahamCampbell @taylorotwell : I just noticed - please check the screenshot from Firefox above and notice domains of ```laravel_session``` cookie. It is not the domain app works with, but rather ```localhost```.

**Comment by GrahamCampbell** [comment]
Is this only happening when you're using environments. Try hard coding all your config.

**Comment by shehi** [comment]
Question: I am not Symfony guru, but as I am currently searching for the cause of ```laravel_session``` cookie being set from ```localhost``` domain in some cases no matter what you do, I stumbled upon ```testClientIpIsAlwaysLocalhostForForwardedRequests()``` unit-test method in Symfony. Question is: is there a chance Symfony resolves the name (our domains, in case of locals and staging areas, our rather FAKE domains which always resolve to 127.0.0.1) to IP address and when it finds out that IP address is rather local, sets ```HTTP=true``` cookies under ```localhost```? It's just a hunch, nothing substantial to back it up right now.

@GrahamCampbell : Ok, I will hard-code configs. I was using ```.env``` file.

**Comment by shehi** [comment]
Meanwhile my ```.env``` file:

![image](https://cloud.githubusercontent.com/assets/653569/7679286/e7bb8634-fd63-11e4-9d3e-b4b1d2c3b542.png)

**Comment by GrahamCampbell** [comment]
> Ok, I will hard-code configs. I was using .env file.

We're already aware of concurrency issues with the environment system. I'm wondering if this session issue is actually just this same issue, applied.

**Comment by Sladewill** [comment]
@shehi Just a word of warning might want to take down that screenshot of your .env file its got your Google and Twitter OAuth details in, might want to obfuscate them before uploading it again.

**Comment by GrahamCampbell** [comment]
@Sladewill Bit late for that. They'll need invalidating.

**Comment by shehi** [comment]
No worries folks, they are for dev purposes, can be invalidated in a whim. :)

**Comment by shehi** [comment]
@GrahamCampbell , I changed to ```"driver" => "file"``` in ```session.php``` config file. Results are the same.

**Comment by GrahamCampbell** [comment]
I'm meaning, don't use the env system at all for any config and see what happens.

**Comment by shehi** [comment]
@GrahamCampbell fixed the problem. Value ```'domain' => ``` inside ```session.php``` file must be set correctly. I set it the same value as my fake real-looking domain - in my case ```basis.audith.org``` - and session is working now. Setting that value as ```localhost``` and accessing the app from ```localhost``` domain does NOT work in Chrome (as per aforementioned reason: cookies are not designed for TLDs and Chrome follows that rule), but works in Firefox. Setting it to ```audith.basis``` (fake domain I used for my project in general) works as well. 

To those who work with Socialite, be warned: since Twitter doesn't accept ```localhost``` and G+ doesn't accept unreal domains like ```audith.basis```  in callback-URIs, you should be working with real-looking domains to avoid all these problems altogether.

**Comment by shehi** [comment]
Or we could just set that line like: ```'domain' => env('SESSION_COOKIE_DOMAIN', 'localhost'),``` and move that setting to dotEnv?

**Comment by TheShiftExchange** [comment]
@shehi - setting the domain to solve your session problem doesnt sound like rest of the problem the rest of us are having. We have random session fails, sometimes under heavy load from AJAX.

But our sessions *work* most of the time. Yours doesnt - so it would seem to be a different issue than discussed here.

**Comment by shehi** [comment]
@TheShiftExchange Yea, I think it might be. That might be related to that "race-condition" topic mentioned before, I guess. Or something related to ```file_get_content()``` not locking file before reading (which would clearly break concurrency in session handling). 

@GrahamCampbell , I believe my posts might have diverted this discussion away from original topic/problem at hand. If you guys believe this is the case, please feel free to delete my posts and put one tiny notification for those who would have the same problem I did today. No need to clutter this topic, if my posts are close to irrelevant.

**Comment by GrahamCampbell** [comment]
> We have random session fails, sometimes under heavy load from AJAX.

I suspect that's caused by the env system. Can anyone confirm this?

**Comment by TheShiftExchange** [comment]
@GrahamCampbell - if people run `config:cache` - would they still be subjected to env system concurrency issue? I'm guessing not?

**Comment by GrahamCampbell** [comment]
> if people run config:cache - do you think they would still be subjected to env system concurrency issue? I'm guessing not?

True. Are people still experiencing this after caching the config?

**Comment by GrahamCampbell** [comment]
That is, with drivers other than file. People were complaining of issues with redis too?

**Comment by TheShiftExchange** [comment]
Ping @ammont @esbenp @Naxiz @Sladewill @ifox

Are you running Laravel 5? If yes - are you using `config:cache`? If not - can you try and use that feature (which is helpful in production anyway) and see if the problem continues?

**Comment by janhartigan** [comment]
@GrahamCampbell that would make sense. If the putenv was being unset for that second request, the Encrypter wouldn't be using the correct app key to decrypt the cookie.

Here are the relevant threads:

https://github.com/vlucas/phpdotenv/issues/76

https://github.com/laravel/framework/commit/866a8cfc58430830069bd24feacf2d81e99385a2#commitcomment-9759653

https://github.com/laravel/framework/issues/7354

https://github.com/laravel/framework/pull/8187

**Comment by Sladewill** [comment]
We aren't using config:cache at all. Not any other caching in our development environments. Although since I switched to redis the issue has stopped happening.

**Comment by GrahamCampbell** [comment]
I think this is pretty much sorted. To avoid this bug, use config caching and a driver that's not "file".

**Comment by GrahamCampbell** [comment]
The "file" driver was never intended for large scale production use.

**Comment by franzliedke** [comment]
It does apparently fail with a few concurrent AJAX requests (on a single page), which has nothing to do with production...

**Comment by GrahamCampbell** [comment]
> It does apparently fail with a few concurrent AJAX requests (on a single page), though, which has nothing to do with production...

What does. Does redis fail after using config:cache first?

**Comment by franzliedke** [comment]
The file driver.

**Comment by GrahamCampbell** [comment]
> The file driver.

We already know that's broken, and it's not getting fixed. That wasn't the purpose of this issue. Look at the very top.

**Comment by GrahamCampbell** [comment]
> still seem to be an issue, even for drivers other than file.

**Comment by GrahamCampbell** [comment]
I opened this to investigate drivers other than the file one because there's no way they should be failing, and that seems to be resolved. Taylor already rejected the overly strict file driver fixes. The recommended advice would be to not use the file driver.

**Comment by Naxiz** [comment]
I discovered the bug using the database driver, so it’s certainly not only the file driver. It seems it doesn’t matter whether is use the file or database driver.

**Comment by janhartigan** [comment]
@Naxiz did you also try caching your config?

**Comment by Naxiz** [comment]
I did, and I got a TokenMismatchException after just a few submits using my own test case. These are not even AJAX or concurrent connections. Unfortunately I don’t have much time to look into this issue myself at the moment, I’ll report back if I find anything useful later.

**Comment by janhartigan** [comment]
@GrahamCampbell so given that it seems to remain unclear whether this is related to dotenv.

**Comment by GrahamCampbell** [comment]
@Naxiz As I said, we're not interested in the file driver. We already know it's limited.

**Comment by Stichoza** [comment]
I'm using **config caching**, but still I have session persistence issues. Session in not being stored sometimes (randomly).

Both on `file` and `redis` drivers.

**Comment by sandrodz** [comment]
This is causing us quite a lot of headache, days of debugging and the problem turns out to be a session handing in laravel :/

**Comment by GrahamCampbell** [comment]
> This is causing us quite a lot of headache, days of debugging and the problem turns out to be a session handing in laravel :/

What driver?

**Comment by sandrodz** [comment]
We tried both `file` and `redis`

**Comment by GrahamCampbell** [comment]
:(

**Comment by Stichoza** [comment]
plus config caching enabled

**Comment by sandrodz** [comment]
To add to the pain, issue is intermittent. So we get 3 invalids in 30 tries.

**Comment by taylorotwell** [comment]
3 invalid what?

**Comment by janhartigan** [comment]
@sandrodz how rapid are requests? Are you making them in quick succession or are they spaced out?

**Comment by sandrodz** [comment]
@Stichoza knows better, I've been following development only from the managers perspective.

@janhartigan  no pattern there really. lets say 5 secs <-> 15 secs. I also noticed that on windows+chrome combination it was easier to reproduce problem than on mac+chrome, but this could have been pure coincidence. Also number of users/sessions seemed to have worsened the problem.

@taylorotwell bad wording on my part, we store specific key in the user session. In rare occasions 3 out of 30 this wasn't set correctly. But as I said @Stichoza knows better, so lets wait for his insight.

**Comment by Naxiz** [comment]
@GrahamCampbell I'm having this issue with the `database` driver, also with the config caching. Also, there is at least 5 seconds between the requests but it still happens.

**Comment by Stichoza** [comment]
The issue appears at nearly no concurrent requests. I mean just me testing on homestead. Session is not being saved sometimes.

Simplified TL;DR: I set session key `last_viewed_post` to a certain value (`id` of post) in `/post/{id}` route. I have also `/get-last-viewed-id` where I output the value of `session('last_viewed_post')`. Sometimes the value is old one, I mean session key isn't being saved.

Issue remained when we switched to redis as a session driver and enabled config caching. The only solution I figured out, is that now I'm saving its' value in `last_viewed_post` **field** on user object and it works fine.

**Comment by janhartigan** [comment]
@Stichoza can you try to create just a simple GET route that you hit repeatedly with a page refresh until you get the problem? I'm trying to determine if the problem is related to multiple concurrent AJAX requests in certain environments.

**Comment by taylorotwell** [comment]
I think we're particularly interested in a demo application that can
recreate the problem in Homestead using the Redis or Memcached driver.

**Comment by Stichoza** [comment]
@taylorotwell I reproduced it and as @janhartigan said, it fails when I rapidly hit "get" route with browser refresh.

#### Demo app
I created a demo app at my local homestead:
 - http://pamparam.localtunnel.me/demo/set
 - http://pamparam.localtunnel.me/demo/get

#### Source code

```php
// routes.php

get('demo/set', function() {
	$rand = rand(1, 10000);
	session(['rand' => $rand]);
	return 'Value set to: ' . $rand;
});

get('demo/get', function() {
	return 'Value is: ' . session('rand', 'N/A');
});
```

#### Example
![screen shot 2015-05-23 at 01 01 38](https://cloud.githubusercontent.com/assets/1139050/7779663/7a500e26-00e7-11e5-839b-4d6b3241378f.png)

#### How to reproduce

Open two tabs as shown on screenshot above. Hit `get` route once to set random number. Then refresh the same page and then rapidly hit `get` route (Click ⌘+R several times quickly). Value remains the same (old) in get route.

I'll try to record a video now.

**Comment by janhartigan** [comment]
@Stichoza so just to clarify...you can't replicate the problem without rapid refreshes?

**Comment by Stichoza** [comment]
@janhartigan I'll tried now (~50 tries) and couldn't. I only replicated when I refreshed `get` quickly.

It seems like the problem occurs when the following happens:

 - started: saving to session. (`set` route)
 - read the value. (`get` route)
 - finished: saving to session. (`set` route)

**Comment by janhartigan** [comment]
@GrahamCampbell looks suspiciously like the multi-thread problem, but I can't imagine how that would matter when config is being cached.

**Comment by Stichoza** [comment]
Here's a video I just recorded: https://youtu.be/_sE1uc90snc

Driver: Redis
Config: cached
Environment: Homestead 2.0.7

**Comment by taylorotwell** [comment]
Trying to figure out what you are showing in that video. Are you trying to
demonstrate at the end how your value gets stuck on that single integer?

**Comment by Stichoza** [comment]
Exactly. Take a closer look at active windows (can be seen using window shadows) and refresh buttons.

 - I refresh `get` route. Prints generated random number and sets it to session.
 - Then I gently refresh `get` route that outputs value from session (see code posted by me above)
 - I repeat this action again and everything seems to be fine.
 - Then I refresh `set` route again, quickly switch to `get` and rapidly hit CMD+R several times.
 - `set` route outputs new random number it set to session.
 - Then I gently refresh `get` route several times with normal intervals and the see the value is the same as it was in step 3.

**Comment by taylorotwell** [comment]
I can't recreate the issue @Stichoza is talking about... I do the cmd + R thing as fast as I can and i see it refreshing, but then if I set a new number to the session, refreshing `get` again gets the new value correctly.

**Comment by Stichoza** [comment]
@taylorotwell You should start rapidly hitting `get` before `set` finishes its' work and outputs new value. I'm doing it at 0:08-0:18 in the video: https://youtu.be/_sE1uc90snc?t=8s

**Comment by taylorotwell** [comment]
Did that too... I can't recreate it.

**Comment by Sladewill** [comment]
I thought this was solved by switching to redis, but today I was randomly getting this issue as well, not even spamming my system, just generally browsing though it and it came up with TokenMismatchException and the session was gone.

**Comment by l3ehnam** [comment]
1) I did open this page on first window http://pamparam.localtunnel.me/demo/set
2) and this one on the other window http://pamparam.localtunnel.me/demo/get
3) pressed cmd + r on the first window for more than 30 times
4) refreshed the second window and the numbers were different

**Comment by Stichoza** [comment]
@l3ehnam I think you're doing different thing. You should refresh `set` to set a new value to session and until it returns the result, you should refresh `get` several times quickly. During your refreshes the `set` will set a new value to session (but will fail silently). You continue refreshing `get` route and the value remains unchanged, i.e. `set` didn't really set the value to session.

/cc @taylorotwell

**Comment by taylorotwell** [comment]
I think this thread is sort of becoming a dumping ground for anyone who has any random issue with the session. :/

**Comment by l3ehnam** [comment]
@Stichoza your right! did the same test this time and got different numbers

**Comment by sandrodz** [comment]
@taylorotwell this is not a random issue, the problem is real, and quite annoying, recreated by number of people here. We can go in two directions from here, pretend that there is no problem and continue with our lives, or actually acknowledge its existence and do something about it.

**Comment by sn0opr** [comment]
Maybe it's a hardware/environment problém ? because I can't reproduce the issue

**Comment by sn0opr** [comment]
@Stichoza I think it's from the service that you are using (localtunnel.me) I think it's caching or somthing like that, try at localhost

**Comment by sandrodz** [comment]
@sn0opr we can reproduce problem on homestead+mac. And ubuntu 12.04 LTS VPS.

**Comment by sn0opr** [comment]
@Stichoza I really tried many times without any problem 
on Windows 8.1 WAMP server

**Comment by taylorotwell** [comment]
Nobody is denying the problem is real. If you are repeatedly changing the
session on each GET request, you're going to possibly get stale data. The
only way around that is to makes the sessions totally blocking, which we
don't want to do, or to tweak your app to not write to sessions on
idempotent requests like GET.

**Comment by Stichoza** [comment]
> If you are repeatedly changing the session on each GET request

No, I'm not changing anything on `get` request. I just read data and output it. Session in changed on `set` request which I hit only once while replicating issue.

**Comment by taylorotwell** [comment]
What do you mean session is changed? If you do a set request but a get
request starts before that request AND then finished after the set, the get
request is going to write the stale data back to the session.

**Comment by Stichoza** [comment]
@taylorotwell "session is changed" I mean value of a key in session is changed.

> the get request is going to write the stale data back to the session

So that's the problem. I'll paste the code @taylorotwell posted in Slack chat yesterday:

> ```php
> Route::get('/', function () {
>     session(['foo' => 'foo']);
> });
> 
> get('get', function() {
>     sleep(10);
>     return session('foo');
> });
> 
> get('set', function() {
>     sleep(5);
>     session(['foo' => 'bar']);
> });
> ```
> 
> all you have to do is hit / to initialize the value to foo. then if you hit /set, then swap over and hit /get before the set finishes, the value will still be foo instead of being updated to bar and if you keep hitting get its never going to change. its just because the get read the session while it was still foo… the set request updated to bar… but the get request still had foo in its session attribute array when it re-wrote the session back and the bar was lost.

So the problem is that session data is being written back at the end on **every** request, whether or not session data was changed.

So the only fix would be to write back session data **only** if session data was changed during current request.

**Note:** The problem will still occur if we re-read session data at the end of the session. It will see data was changed - it will compare newly read data to one that we read at the beginning, so it will still write back stale data.

We should write back session data, **only** if `Session::put()` (or any other method that changes session data) was called.

As @taylorotwell says, it will make the expiration stuff complicated. So at this moment I don't know how to avoid this problem.

**Comment by radmen** [comment]
> We should write back session data, only if Session::put() (or any other method that changes session data) was called.
> As @taylorotwell says, it will make the expiration stuff complicated. So at this moment I don't know how to avoid this problem.

I understand that expiration stuff can become a problem when session is saved only on change. For year's (before Laravel) I've been using this solution in our company framework. It was used on pages with big traffic, and we didn't experience any problem with sessions being expired due not saving any data to it.

I think that this is a risk worth taking.

**Comment by taylorotwell** [comment]
It is sort of a problem though. If I log into the site and I work in the
app all day. I don't want to be logged out every two hours just because my
session hasn't changed. That's silly.

**Comment by barryvdh** [comment]
And what about the [touch()](http://php.net/manual/en/function.touch.php) function? Would that still corrupt the file? It would prevent the session going stale, right?

	if ($dataChanged) {
		writeSession();
	} else {
		touchSession();
	}

Or take the time into consideration, only update once a minute?

	if ($dataChanged || $secsAgoChanged > 60) {
		writeSession();
	}

Or combine both?

	if ($dataChanged ) {
		writeSession();
	} elseif ($secsAgoChanged > 60) {
		touchSession();
	}

**Comment by taylorotwell** [comment]
The seconds thing will be error prone. To me it makes more sense to just
evaluate why is the session changing in s particular app and perhaps
structure your AJAX calls to where they aren't going to be going in between
set requests for the same session. Perhaps use a cookie to store a piece of
data, etc

**Comment by Stichoza** [comment]
@taylorotwell What about long-polling? It will screw up thing if someone has long-polling implemented in app.

**Comment by taylorotwell** [comment]
What will screw up things?

**Comment by Stichoza** [comment]
This issue. I mean if you change anything in session data, long-polling requests will overwrite stale data to session as each polling request will take up to 40 seconds.

**Comment by taylorotwell** [comment]
I mean there are several options there. You could send a custom header,
check for that header, and then override the session middleware. That's
just off the top of my head. I'm not sure of a really great solution that
works well in every situation.

**Comment by janhartigan** [comment]
Also worth mentioning that the near-term future of bidirectional communication likely won't include long-polling unless your app needs to support IE9.

**Comment by sitesense** [comment]
I'm not sure I fully understand the "expiration" problem, but if it's a case that the session must save at certain intervals to avoid being logged out, how about using the garbage collection/lottery approach to save the session occasionally? Or maybe just add a session save in the "gc" function of the session handler?

**Comment by sn0opr** [comment]
@taylorotwell when it comes to long-polling it's a big problem, I used to store data from Session data while there is a long-polling ajax request, the Session forget even in the same request, so we suitched to cookies so we didn't faced this problem anymore.

**Comment by suncoastkid** [comment]
Same issue here running L5 on PHP 5.6.9. - TokenMismatchException VerifyCsrfToken.php on line 46 - on every post request. Works fine on PHP 5.4.

**Comment by sandrodz** [comment]
We are running it on PHP 5.5.24-1

**Comment by shehi** [comment]
@suncoastkid : Yea, it happens because Session is re-initialized on every request (make sure you have set the domain-name setting in ```config/session.php``` btw).

**Comment by DeBaas** [comment]
We have exactly the same problem. Weird thing is we can not reproduce it ourselves but we have Bugsnag installed and see the tokenmismatch exception pop-up twice per minute or so, completely random. Random users, random computers, random browsers. Only thing in common is that it where all ajax calls. Seems like about 0,1% of all ajax calls return a token mismatch exception, all the ajax calls suffer from this. I know that it has nothing to do with a session expiring because we have one ajax call which is only loaded first thing when a page loads and even this ajax call will return a tokenmismatch exception about 0,1% of the time, again completely random but here at our office we can not reproduce it.
If somebody wants access to the Bugsnag report, I can give you that for investigation purposes.

Environment: AWS Beanstalk (64bit Amazon Linux 2015.03 v1.4.1 running PHP 5.6)
Latest Laravel 5.0.x
Website: www.studocu.com

We used to have 1 AWS redis node for both the cache and sessions. Today we tried separating the 2 and moved the session part to a newly setup memcached node, but this didn't make a difference and the problem still exists.

**Comment by nathanburgess** [comment]
Experiencing this issue with a fresh install of Laravel 5 on a recently provisioned Ubuntu 15.04 server. No matter what route, it's always a token mismatch. Tried both file and database session drivers. I'd be willing to let @GrahamCampbell or @taylorotwell or someone else into the server to poke around and see if they can figure it out.

**Comment by TheShiftExchange** [comment]
@nathanburgess -You have a complete session failure - which is a different issue to this thread. Please post that issue in a forum for assistance.

**Comment by GrahamCampbell** [comment]
Can people stop abusing this thread and actually read it, and only comment if they have something useful to say.

**Comment by DeBaas** [comment]
@GrahamCampbell Was this comment meant for me? If so, how can I provide more information?

**Comment by shehi** [comment]
@DeBaas Open independent issue for your case, because it is not related to this one (this one = random, unpredictable and untrackable session failures [0.1-0.3% occurrence rate]).

**Comment by DeBaas** [comment]
@shehi Did you even read my report? I'll quote myself "Seems like about 0,1% of all ajax calls return a token mismatch exception, all the different ajax calls suffer from this. Random users, random computers, random browsers".

This is caused by a session failure and thus a session regeneration causing random (0,1% of all) ajax calls to return a tokenmismatch.

**Comment by TheShiftExchange** [comment]
@DeBaas - no - the comment was not meant for you. It was for another comment that has since been deleted.

**Comment by shehi** [comment]
And we are filling this topic with more nonsense. I am deleting my posts in a bit...

**Comment by axlf** [comment]
I had a problem with persisting sessions. (Not sure if its related to this problem). The first time I installed Laravel via installer, then I had problems running phpunit and the weird session persisting problem.
Then I reinstalled Laravel via composer and now everything works as expected.

Perhaps its a bug in the installer.

**Comment by ajeto** [comment]
I had a problem with session persistence only on live server, but I solved it. The problem was in blank space after php closing tag in my Profile model which I using  for authentication. After I removed it everything is work now.

**Comment by GrahamCampbell** [comment]
> I had a problem with session persistence only on live server, but I solved it. The problem was in blank space after php closing tag in my Profile model which I using for authentication. After I removed it everything is work now.

That is totally unrelated! Also, you should not use php closing tags for that reason.

**Comment by sameerpanjwani** [comment]
Been following this thread for a while hoping for a solution. I just bumped into a situation which _might be_ related to this. (Using the *database* session driver)

Accidentally I put a wrong route for a form and when I submitted I obviously got a ````NotFoundHttpException in RouteCollection.php````. I anyway submitted the form again but this time got a *TokenMismatch". I corrected the route and then submitted and it worked.

Now why I think it's related is because I have noticed from time to time a ````NotFoundHttpException```` even on valid routes from time to time on fast AJAX requests or situations where there been many concurrent requests. And after that I've seen this ````TokenMismatchException````

Now this a speculative guess but could this TokenMismatch exception be somehow related whenever a ```NotFoundHttpException```` or some other exceptions are thrown by the Laravel application which results in the session for some reason getting lost. I might be sounding really ignorant out here by even giving this reasoning but I just thought it might be a possibility. 

Also, any reason why I would in any case get a *TokenMismatch* the second time on submitting a form to an invalid route instead of a Not Found page. Is that expected behavior?

**Comment by TheShiftExchange** [comment]
> Also, any reason why I would in any case get a TokenMismatch the second time on submitting a form to an invalid route instead of a Not Found page. Is that expected behavior?

CSRF token checks run as part of the global middleware, before you reach the routes. So an invalid CSRF will trigger an exception, regardless of the url.

**Comment by sameerpanjwani** [comment]
Yeah I got that, just didn't understand why a csrf exception was triggered after a not found exception the first time.

**Comment by sameerpanjwani** [comment]
Ok figured that whenever hit a ````NotFoundHttpException```` then the session is emptied and the _token_ is regenerated. So that's why when I hit the back button, it had the old token and I had to refresh the page to get the new token. Same is the case with Ajax requests, whenever for some reason the ````NotFoundHttpException``` is triggered (which happens weirdly on rare occasions even on valid routes - _guess that's the bug?_) then also the ```TokenMismatchException```` is triggered. Is it necessary to clear the session / regenerate the token if a route is not found?

**Comment by sameerpanjwani** [comment]
Ok figured that whenever hit a ````NotFoundHttpException```` then the session is emptied and the _token_ is regenerated. So that's why when I hit the back button, it had the old token and I had to refresh the page to get the new token. Same is the case with Ajax requests, whenever for some reason the ````NotFoundHttpException``` is triggered (which happens weirdly on rare occasions even on valid routes - _guess that's the bug?_) then also the ```TokenMismatchException```` is triggered. Is it necessary to clear the session / regenerate the token if a route is not found?

**Comment by Apstreet** [comment]
Hi!
I have the same problem. The session of my app expired when I do new request. In a local environment that's fell good but in a real hosting the session expires randomly, someone have some hint?

**Comment by OskarD** [comment]
I started getting this today after several days of development.

Windows 8.1
Google Chrome Version 43.0.2357.124 m
PHP Version 5.6.8

Notable information:
* Only happens on my Chrome browser (Tested latest IE and latest FF)
* It happens in incognito mode as well
* I get the error more often than not (About 90% of the time I submit a form)
* It does not appear to have anything to do with the frequency of form submissions
* It happens in both prod and local APP_ENV

**Comment by diegofelix** [comment]
I have the same problem and here is how i'm tested and fixed the problem.

First, i knew i have some 404 ajax and resources ( javascripts ) on my site. So I logged on the site, and put four tabs opened with my site in firefox.

The Firefox has a nice tool called "Reload all tabs". So I did it and with a few refresh all tabs I get logout.

Then, I removed all 404 requests and tried again, and, after 20 attempts, no logouts from now.

**Comment by Kuijkens** [comment]
Guys, I work at the same company as @DeBaas .
All of your problems are not related to this issue! 

@OskarD You are either not using the Laravel Forms which will include the _token or you are doing something else wrong! This issue is about random session regeneration, not by 90%+ of your forms returning 'the error'!

@diegofelix @sameerpanjwani  Obviously 404 response will return a tokenmismatch exception afterwards to other requests or calls! If you got 1 error somewhere, your application will die and therefore refresh your session. All other requests will then return a tokenmismatch exception.

Just to clarify: If you do not have a single error or an other header response then 200 and still about 0,1% of your (ajax) requests are returning a tokenmismatch exception due to a session regeneration (check this first in your logs!), then you can comment in this issue thread.

**Comment by OskarD** [comment]
@Kuijkens I don't know why it works sometimes, but when I get 'the error' it's because the form is sending a different (Seemingly random) token than what is stored in the session. Are you sure this is not the same issue?

**Comment by sameerpanjwani** [comment]
@Kuijkens so it's expected behaviour for session and token to get refreshed after a 404? Any reason for that being necessary or that just the way it's made?

And some times a 404 is triggered even on valid routes? That's what seems to happen in my environment randomly...it's actually first a "Whoops something went wrong...notfoundexception" followed by tokenmismatch...on fast ajax requests...and hence guessed that this could be issue others are facing..and just wanted a way to ensure that even if 404 did come about, the token does not get refreshed...is there any way to ensure that happens then my problem will be solved at least ;) because I can't figure why I get a notfound sometimes on valid routes, it's rare but happens and then it results in this tokenmismatch....

**Comment by Kuijkens** [comment]
@OskarD Don't turn this around on me. You should investigate. Log the session id's, compare them. Check you POST values. Log the different session token's which are being checked. If your problem is that, without a single error anywhere, without a session timeout, your findings are that the session is regenerating about 0,1% of the time without a probable cause, you can report here. Otherwise it is not this issue and you should ask on other sites/forums what to do.

Same goes for @sameerpanjwani , I do not know why you are getting random 404 responses on your various existing routes. That is a totally different issue, probably not a bug but an error in your coding/routing. If you get an exception, doesnt matter what kind of, the app will die and regenerate your session => next ajax calls will all return a tokenmismatch untill you refresh your page with an error.

All of your problems have nothing to do with Random Session regeneration on about 0,1% of the calls without any errors at all!

Im deleting my post in 5 minutes, I suggest you do the same and ask for help elsewhere

**Comment by levacic** [comment]
@Kuijkens

> @diegofelix @sameerpanjwani Obviously 404 response will return a tokenmismatch exception afterwards to other requests or calls! If you got 1 error somewhere, your application will die and therefore refresh your session. All other requests will then return a tokenmismatch exception.

What? Why would your session get refreshed after a 404? This is not true, and I've just confirmed in a clean install.

Also, that would be a horrible user experience, because users would basically be logged out after visiting a missing page.

**Comment by Kuijkens** [comment]
@levacic Sorry my bad, was getting frustrated about everybody just commenting their issues in here. Of-course a 4xx will not refresh your session.

I do not have any reponses other then 200, I do not have concurrent and high frequency ajax calls. No exceptions are being thrown (we log everything on bugsnag). Still once in a while, about 0,1%, a request is returning a tokenmismatch caused by a regeneration of the session (i've actually logged this), not caused by session expiring or any thing like this. We have this problem both on Redis and Memcache. This is actually influencing the user experience on our website!

Since this is only happening on our production environment (we can not reproduce the problem on develop or staging). I had to exclude all vital ajax routes from CSRF verification. Since nobody is interested in our bugsnag report or logs it kinda feels like nobody is working on this issue and people are only cluttering this thread with non related issues. This problem is actually costing us users and thus money and is kind of unacceptable. (it is the only laravel issue we are having btw, everything else works like a charm and is awesome!)

**Comment by TheShiftExchange** [comment]
@Kuijkens - what *exactly* is your production environment config?

**Comment by Kuijkens** [comment]
@TheShiftExchange 

Environment: AWS Beanstalk (64bit Amazon Linux 2015.03 v1.4.1 running PHP 5.6)
PHP 5.6.8
Apache 2.4.12
Latest Laravel 5.0.x
Website: www.studocu.com

**Comment by TheShiftExchange** [comment]
@Kuijkens - so here is an idea; Can you 'duplicate' that AWS environment? Put a copy of your app on it - and confirm the bug still occurs?

Then - if you are trusting enough - could you give @taylorotwell access and the steps to reproduce (if he is open to the idea - I'm just putting it out there).

The biggest issue is the difficulty in replicating the issue. So far we've not been able to get access to an environment where we can duplicate it. I've tried on different cloud providers - and I cant seem to get it to trigger...

**Comment by Kuijkens** [comment]
@TheShiftExchange - I willing to give it a go, but I don't think it will work. Our staging environment is a duplicate of our production. Like I said before, we can not reproduce the bug ourselves (on both production or staging). The only thing that makes this bug visible is a high user volume website and Bugsnag. I'm guessing more applications suffer from this but they don't even know it. It is only happening about once or twice per hour (random calls, random routes, random users) versus thousands of successful requests.

**Comment by TheShiftExchange** [comment]
Can you hit your site with a load tester - like https://www.blitz.io/ - can that trigger it?

**Comment by Kuijkens** [comment]
@TheShiftExchange I doubt that, all our ajax requests are user induced. I should build in a custom ajax request then, which basically does nothing but will load on various times.

**Comment by red55der** [comment]
Same problem here. Randomly. TokenMismatchException line 46. Happens with IE Win8.1 or IOS.
Can't reproduce it.
Simple form application (html+bootstrap+jqueryui) with default Laravel auth
Session store in database
Laravel 5.1
PHP 5.5.9
Apache 2.4
Cleared cache etc

Session cookie gets reinit every request. Sow GET login page has other session then POST login page.

**Comment by dakira** [comment]
For me this problem went away when I updated my Homestead machine image from 0.2.6 to 0.2.7, destroyed my machine and recreated it.

**Comment by Apstreet** [comment]
I solved my problem putting session_start() in the top of my routes file.

**Comment by levacic** [comment]
@Kuijkens I understand your frustration, just wanted to correct the wrong info, as it will only further confuse users reading this thread. I'm subscribed to the issue for the same reason as you - random token mismatch exceptions that cannot be reliably reproduced, though in a Laravel 4.2 app. I'm not posting extra info about my situation since it was mentioned that this issue is specifically concerned with Laravel 5, but I'm watching in case someone figures out the problem, because there's a good chance that the root cause might be similar.

**Comment by GrahamCampbell** [comment]
> session_start() in the top of my routes file.

Oh wow. NEVER do that!

**Comment by Apstreet** [comment]
why? That solve my problem

**Comment by GrahamCampbell** [comment]
Because that's totally incorrect.

**Comment by GrahamCampbell** [comment]
Closing this as zero progress is being made.

**Comment by Apstreet** [comment]
Yes, I know but my desperation  is enough

**Comment by dakira** [comment]
@GrahamCampbell most people seem to experience this with Homestead. Whatever the issue is, updating homestead solved it on all my machines.

BTW: I could re-produce this with the most simple imaginable app that subits a form (no ajax).

**Comment by chaucerbao** [comment]
Not sure how related this is, but I've been playing around with Lumen and with the new 5.1 release, I started migrating and noticed that my sessions kept regenerating on each request.  Tests work fine, but in browser, I couldn't keep a session.  I didn't have this problem w/ the Lumen 5.0 release.

For me, it's not random as it is for some people here, it just doesn't keep the session. I updated Homestead to 0.2.7 hoping it would fix the issue, but no luck.

I think Lumen is using Laravel's session class, hence posting here.

https://github.com/chaucerbao/lumen-starter-kit/tree/feature/lumen-5.1

**Comment by red55der** [comment]
I log the Exception header which show that de TokenMismatch mainly occurs on IE Win8.1 or IOS Safari.

I can reproduce it in IE11 when i set the Privacy on High.

solution that works here:
add a customheader in the middleware stack by create a file P3PHeader in App\Http\Middleware;
```
<?php namespace App\Http\Middleware;

use Closure;

class P3PHeader
{
	public function handle($request, Closure $next)
	{
		$response = $next($request);
		$response->headers->set('P3P', 'CP="IDC DSP COR ADM DEVi TAIi PSA PSD IVAi IVDi CONi HIS"');
		return $response;
	}
}
```

And added 'App\Http\Middleware\P3PHeader' to $middleware in Kernel.php

**Comment by tuurbo** [comment]
This is only an issue on my local machine and after running ```artisan optimize --force``` i no longer see the issue

**Comment by Kuijkens** [comment]
@red55der This is looking very promising! For the first time I'm able to reproduce the error myself. Adding the headers like you suggested solved it for me, at least locally. I'm going to hotfix deploy this right away to production and monitor it for a couple of hours. Will be back with the results asap.

**Comment by mga242** [comment]
@red55der Make sure you exclude responses that are redirects, otherwise it will throw an error: http://stackoverflow.com/questions/30490821/laravel-5-tokenmismatchexception-on-php-5-6-9

**Comment by Kuijkens** [comment]
@red55der @tuurbo Tried both suggestions, random session regenerations still occurs.

We are going to upgrade to L5.1 really soon, hopefully things will magically sort itself out.

**Comment by tuurbo** [comment]
@Kuijkens Do you use apache on your local machine? I run windows 8.1 and just switched to nginx and that also solved the issue for me, just like ```artisan optimize --force``` did while using apache.

**Comment by GrahamCampbell** [comment]
Please continue this on the forums.
